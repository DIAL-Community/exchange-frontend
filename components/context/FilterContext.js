import { createContext, useEffect, useState } from 'react'
import { useRouter } from 'next/router'

const FilterContext = createContext()
const FilterDispatchContext = createContext()

const FilterProvider = ({ children }) => {
  const [search, setSearch] = useState('')
  // Task tracker context only
  const [showFailedOnly, setShowFailedOnly] = useState(false)
  const [showGovStackOnly, setShowGovStackOnly] = useState(false)
  const [showMature, setShowMature] = useState(false)

  const [categoryTypes, setCategoryTypes] = useState([])
  const [sdgs, setSdgs] = useState([])
  const [useCases, setUseCases] = useState([])
  const [workflows, setWorkflows] = useState([])

  const [origins, setOrigins] = useState([])
  const [countries, setCountries] = useState([])
  const [sectors, setSectors] = useState([])
  const [organizations, setOrganizations] = useState([])
  const [tags, setTags] = useState([])
  const [datasetTypes, setDatasetTypes] = useState([])

  const [aggregators, setAggregators] = useState([])
  const [operators, setOperators] = useState([])
  const [services, setServices] = useState([])

  const [years, setYears] = useState([])

  const [products, setProducts] = useState([])

  const [buildingBlocks, setBuildingBlocks] = useState([])
  const [showClosed, setShowClosed] = useState(false)

  const [aggregator, setAggregator] = useState(false)
  const [endorser, setEndorser] = useState(false)
  const [endorserLevel, setEndorserLevel] = useState('')

  const [specialties, setSpecialties] = useState([])
  const [certifications, setCertifications] = useState([])

  const [isEndorsed, setIsEndorsed] = useState(false)
  const [endorsers, setEndorsers] = useState([])
  const [licenseTypes, setLicenseTypes] = useState([])
  const [isLinkedWithDpi, setIsLinkedWithDpi] = useState(false)
  const [showDpgaOnly, setShowDpgaOnly] = useState(false)

  const [softwareCategories, setSoftwareCategories] = useState([])
  const [softwareFeatures, setSoftwareFeatures] = useState([])
  const [comparedProducts, setComparedProducts] = useState([])
  const [productStage, setProductStage] = useState(null)

  const [showBeta, setShowBeta] = useState(false)

  const router = useRouter()
  useEffect(() => {
    const handleRouteChange = (url) => {
      // Transitioning to other pages (pathname changing).
      if (url.indexOf(router.pathname) < 0) {
        setSearch('')
        setShowFailedOnly(false)
        setShowGovStackOnly(false)
        setShowMature(false)

        setCategoryTypes([])
        setSdgs([])
        setUseCases([])
        setWorkflows([])

        setOrigins([])
        setCountries([])
        setSectors([])
        setOrganizations([])
        setTags([])
        setDatasetTypes([])

        setAggregators([])
        setOperators([])
        setServices([])

        setYears([])

        setProducts([])

        setBuildingBlocks([])
        setShowClosed(false)

        setAggregator(false)
        setEndorser(false)
        setEndorserLevel('')

        setSpecialties([])
        setCertifications([])

        setIsEndorsed(false)
        setEndorsers([])
        setLicenseTypes([])
        setIsLinkedWithDpi(false)
        setShowDpgaOnly(false)

        setSoftwareCategories([])
        setSoftwareFeatures([])
        setComparedProducts([])
        setProductStage('')

        setShowBeta(false)
      }
    }

    router.events.on('routeChangeComplete', handleRouteChange)

    return () => {
      router.events.off('routeChangeComplete', handleRouteChange)
    }
  }, [router.events, router.asPath, router.pathname])

  const valueProps = {
    search,

    showFailedOnly,
    showGovStackOnly,
    showMature,
    categoryTypes,
    sdgs,
    useCases,
    workflows,

    countries,
    datasetTypes,
    organizations,
    origins,
    sectors,
    tags,

    aggregators,
    operators,
    services,

    years,

    products,

    buildingBlocks,
    showClosed,

    aggregator,
    endorser,
    endorserLevel,
    specialties,
    certifications,

    isEndorsed,
    endorsers,
    licenseTypes,
    isLinkedWithDpi,
    showDpgaOnly,
    softwareCategories,
    softwareFeatures,
    comparedProducts,
    productStage,

    showBeta
  }

  const dispatchProps = {
    setSearch,
    setShowFailedOnly,
    setShowGovStackOnly,
    setShowMature,
    setCategoryTypes,
    setSdgs,
    setUseCases,
    setWorkflows,

    setCountries,
    setDatasetTypes,
    setOrganizations,
    setOrigins,
    setSectors,
    setTags,

    setAggregators,
    setOperators,
    setServices,

    setYears,

    setProducts,

    setBuildingBlocks,
    setShowClosed,

    setAggregator,
    setEndorser,
    setEndorserLevel,
    setSpecialties,
    setCertifications,

    setIsEndorsed,
    setEndorsers,
    setLicenseTypes,
    setIsLinkedWithDpi,
    setShowDpgaOnly,
    setSoftwareCategories,
    setSoftwareFeatures,
    setComparedProducts,
    setProductStage,

    setShowBeta
  }

  return (
    <FilterContext.Provider value={{ ...valueProps }}>
      <FilterDispatchContext.Provider value={{ ...dispatchProps }}>
        {children}
      </FilterDispatchContext.Provider>
    </FilterContext.Provider>
  )
}

export { FilterProvider, FilterContext, FilterDispatchContext }
