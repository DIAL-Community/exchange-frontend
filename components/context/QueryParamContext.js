import { createContext, useState } from 'react'

const QueryParamContext = createContext()
const QueryParamContextProvider = ({ children }) => {
  const [interactionDetected, setInteractionDetected] = useState(false)
  return (
    <QueryParamContext.Provider value={{ ...{ interactionDetected, setInteractionDetected } }}>
      {children}
    </QueryParamContext.Provider>
  )
}

export { QueryParamContextProvider, QueryParamContext }
