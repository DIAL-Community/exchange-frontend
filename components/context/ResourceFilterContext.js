import React, { createContext, useState } from 'react'

const ResourceFilterContext = createContext()
const ResourceFilterDispatchContext = createContext()

const ResourceFilterProvider = ({ children }) => {
  const [search, setSearch] = useState('')
  const [resourceTypes, setResourceTypes] = useState([])
  const [resourceTopics, setResourceTopics] = useState([])
  const [resourceCountries, setResourceCountries] = useState([])

  const resourceFilterValues = {
    search,
    resourceTypes,
    resourceTopics,
    resourceCountries
  }
  const resourceFilterDispatchValues = {
    setSearch,
    setResourceTypes,
    setResourceTopics,
    setResourceCountries
  }

  return (
    <ResourceFilterContext.Provider value={resourceFilterValues}>
      <ResourceFilterDispatchContext.Provider value={resourceFilterDispatchValues}>
        {children}
      </ResourceFilterDispatchContext.Provider>
    </ResourceFilterContext.Provider>
  )
}

export {
  ResourceFilterProvider,
  ResourceFilterContext,
  ResourceFilterDispatchContext
}
