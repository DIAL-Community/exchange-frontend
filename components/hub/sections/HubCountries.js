import { FormattedMessage } from 'react-intl'
import HubCountryTile from '../fragments/HubCountryTile'

const HubCountries = () => {

  return (
    <div className='flex flex-col gap-6 pb-12'>
      <img className='h-32 xl:h-80 w-full object-cover' alt='DIAL DPI Resource Hub' src='/images/hero-image/hub-hero.svg' />
      <div className='absolute w-full left-1/2 -translate-x-1/2 min-h-[20rem]' style={{ top: 'var(--ui-header-height)' }}>
        <div className='flex flex-col gap-8 justify-center mx-auto py-12 px-4 xl:flex-row xl:py-20'>
          <div className='md:text-xl 2xl:text-2xl text-center text-dial-cotton lg:max-w-prose line-clamp-6'>
            <FormattedMessage
              id='dpi.country.subtitle'
              values={{
                break: () => <br />
              }}
            />
          </div>
        </div>
      </div>
      <div className='px-4 lg:px-8 xl:px-56  min-h-[40vh] 2xl:min-h-[50vh]'>
        <div className='flex flex-col gap-6'>
          <HubCountryTile />
        </div>
      </div>
    </div>
  )
}

export default HubCountries
