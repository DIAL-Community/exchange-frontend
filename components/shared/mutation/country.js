import { gql } from '@apollo/client'

export const DELETE_COUNTRY = gql`
  mutation DeleteCountry($id: ID!) {
    deleteCountry(id: $id) {
      country {
        id
        slug
        name
      }
      errors
    }
  }
`

export const CREATE_COUNTRY = gql`
  mutation CreateCountry(
    $name: String!
    $slug: String!
    $description: String
  ) {
    createCountry(
      name: $name
      slug: $slug
      description: $description
    ) {
        country {
          id
          name
          slug
          code
          codeLonger
          description
        }
        errors
      }
    }
`
