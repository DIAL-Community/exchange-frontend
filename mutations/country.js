import { gql } from '@apollo/client'

export const DELETE_COUNTRY = gql`
  mutation DeleteCountry(
    $id: ID!
  ) {
    deleteCountry(
      id: $id
    ) {
      country {
       id
       slug
       name
      }
    }
  }
`

export const CREATE_COUNTRY = gql`
  mutation CreateCountry(
    $name: String!
    $slug: String!
    $code: String
    $codeLonger: String
  ) {
    createCountry(
      name: $name
      slug: $slug
      code: $code
      codeLonger: $codeLonger
    ) {
        country {
          slug
        }
        errors
      }
    }
`
