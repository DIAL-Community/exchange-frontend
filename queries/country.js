import { gql } from '@apollo/client'

export const COUNTRY_SEARCH_QUERY = gql`
  query Countries($search: String!) {
    countries(search: $search) {
      id
      name
      slug
    }
  }
`

export const COUNTRY_CODES_QUERY = gql`
  query Countries($search: String!) {
    countries(search: $search) {
      codeLonger
    }
  }
`

export const COUNTRIES_LIST_QUERY = gql`
  query SearchCountries(
    $first: Int
    $after: String
    $search: String
  ) {
    searchCountries(
      first: $first
      after: $after
      search: $search
    ) {
      totalCount
      pageInfo {
        endCursor
        startCursor
        hasPreviousPage
        hasNextPage
      }
      nodes {
        id
        name
        slug
      }
    }
  }
`
