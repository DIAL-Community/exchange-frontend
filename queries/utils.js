const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMature = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input,
    mature: true
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}
