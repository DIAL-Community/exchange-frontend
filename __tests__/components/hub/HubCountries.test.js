import { screen } from '@testing-library/dom'
import { FilterContextProvider } from '../../../components/context/FilterContext'
import { QueryParamContextProvider } from '../../../components/context/QueryParamContext'
import { ResourceFilterProvider } from '../../../components/context/ResourceFilterContext'
import HubCountries from '../../../components/hub/sections/HubCountries'
import { COUNTRIES_WITH_RESOURCES_SEARCH_QUERY } from '../../../components/shared/query/country'
import { render } from '../../test-utils'
import CustomMockedProvider, { generateMockApolloData } from '../../utils/CustomMockedProvider'
import { mockNextUseRouter, mockTenantApi } from '../../utils/nextMockImplementation'
import { countriesWithResources } from './data/HubCountries.data'

mockTenantApi()
mockNextUseRouter()
describe('Unit tests for the opportunity detail page.', () => {
  const mockCountriesWithResources = generateMockApolloData(
    COUNTRIES_WITH_RESOURCES_SEARCH_QUERY,
    {
      'search': ''
    },
    null,
    countriesWithResources
  )

  test('Should render detail of a opportunity.', async () => {
    const { container } = render(
      <CustomMockedProvider mocks={[mockCountriesWithResources]}>
        <QueryParamContextProvider>
          <FilterContextProvider>
            <ResourceFilterProvider>
              <HubCountries />
            </ResourceFilterProvider>
          </FilterContextProvider>
        </QueryParamContextProvider>
      </CustomMockedProvider>
    )

    expect(await screen.findByText('Zambia')).toBeInTheDocument()
    expect(await screen.findByText('Uganda')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
