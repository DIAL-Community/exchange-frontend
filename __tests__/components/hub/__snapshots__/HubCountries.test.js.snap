// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for the opportunity detail page. Should render detail of a opportunity. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="flex flex-col gap-6 pb-12"
  >
    <img
      alt="DIAL DPI Resource Hub"
      class="h-32 xl:h-80 w-full object-cover"
      src="/images/hero-image/hub-hero.svg"
    />
    <div
      class="absolute w-full left-1/2 -translate-x-1/2 min-h-[20rem]"
    >
      <div
        class="flex flex-col gap-8 justify-center mx-auto py-12 px-4 xl:flex-row xl:py-20"
      >
        <div
          class="md:text-xl 2xl:text-2xl text-center text-dial-cotton lg:max-w-prose line-clamp-6"
        >
          
    Find tools and resources related to how DPI has been implemented in a specific country.
  
        </div>
      </div>
    </div>
    <div
      class="px-4 lg:px-8 xl:px-56  min-h-[40vh] 2xl:min-h-[50vh]"
    >
      <div
        class="flex flex-col gap-6"
      >
        <div
          class="grid grid-cols-1 md:grid-cols-2 xl:grid-cols-3 gap-3"
        >
          <a
            href="/hub/countries/et"
          >
            <div
              class="border border-rounded bg-dial-slate-100 flex gap-4 items-center"
            >
              <img
                alt="Flag for country: ET."
                class="object-contain w-20 h-20 px-4"
                src="https://flagsapi.com/ET/flat/64.png"
              />
              <div
                class="font-semibold"
              >
                Ethiopia
              </div>
            </div>
          </a>
          <a
            href="/hub/countries/ke"
          >
            <div
              class="border border-rounded bg-dial-slate-100 flex gap-4 items-center"
            >
              <img
                alt="Flag for country: KE."
                class="object-contain w-20 h-20 px-4"
                src="https://flagsapi.com/KE/flat/64.png"
              />
              <div
                class="font-semibold"
              >
                Kenya
              </div>
            </div>
          </a>
          <a
            href="/hub/countries/ug"
          >
            <div
              class="border border-rounded bg-dial-slate-100 flex gap-4 items-center"
            >
              <img
                alt="Flag for country: UG."
                class="object-contain w-20 h-20 px-4"
                src="https://flagsapi.com/UG/flat/64.png"
              />
              <div
                class="font-semibold"
              >
                Uganda
              </div>
            </div>
          </a>
          <a
            href="/hub/countries/zm"
          >
            <div
              class="border border-rounded bg-dial-slate-100 flex gap-4 items-center"
            >
              <img
                alt="Flag for country: ZM."
                class="object-contain w-20 h-20 px-4"
                src="https://flagsapi.com/ZM/flat/64.png"
              />
              <div
                class="font-semibold"
              >
                Zambia
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
